src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/SolutionExplorerInProcess.cs

```diff
@@ -0,0 +1,560 @@
+            ErrorHandler.ThrowOnFailure(solution.AddNewProjectFromTemplate(projectTemplatePath, null, null, projectPath, projectName, null, out _));
+        }
+
+        public async Task RestoreNuGetPackagesAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var dte = await GetRequiredGlobalServiceAsync<SDTE, EnvDTE.DTE>(cancellationToken);
+            var solution = (EnvDTE80.Solution2)dte.Solution;
+            foreach (var project in solution.Projects.OfType<EnvDTE.Project>())
+            {
+                await RestoreNuGetPackagesAsync(project.FullName, cancellationToken);
+            }
+        }
+
+        public async Task RestoreNuGetPackagesAsync(string projectName, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var operationProgressStatus = await GetRequiredGlobalServiceAsync<SVsOperationProgress, IVsOperationProgressStatusService>(cancellationToken);
+            var stageStatus = operationProgressStatus.GetStageStatus(CommonOperationProgressStageIds.Intellisense);
+            await stageStatus.WaitForCompletionAsync().WithCancellation(cancellationToken);
+
+            var solutionRestoreService = await GetComponentModelServiceAsync<IVsSolutionRestoreService>(cancellationToken);
+            await solutionRestoreService.CurrentRestoreOperation;
+
+            var projectFullPath = (await GetProjectAsync(projectName, cancellationToken)).FullName;
+            var solutionRestoreStatusProvider = await GetComponentModelServiceAsync<IVsSolutionRestoreStatusProvider>(cancellationToken);
+            if (await solutionRestoreStatusProvider.IsRestoreCompleteAsync(cancellationToken))
+            {
+                return;
+            }
+
+            var solutionRestoreService2 = (IVsSolutionRestoreService2)solutionRestoreService;
+            await solutionRestoreService2.NominateProjectAsync(projectFullPath, cancellationToken);
+
+            while (true)
+            {
+                if (await solutionRestoreStatusProvider.IsRestoreCompleteAsync(cancellationToken))
+                {
+                    return;
+                }
+
+                await Task.Delay(TimeSpan.FromMilliseconds(50), cancellationToken);
+            }
+        }
+
+        public async Task OpenFileAsync(string projectName, string relativeFilePath, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var filePath = await GetAbsolutePathForProjectRelativeFilePathAsync(projectName, relativeFilePath, cancellationToken);
+            VsShellUtilities.OpenDocument(ServiceProvider.GlobalProvider, filePath, VSConstants.LOGVIEWID.Code_guid, out _, out _, out _, out var view);
```

Comment:
iirc there is an async document open service now - https://devdiv.visualstudio.com/DevDiv/_git/VS.RPC.Contracts?path=%2Fsrc%2FMicrosoft.VisualStudio.RpcContracts%2FOpenDocument%2FIOpenDocumentService.cs&_a=contents&version=GBmain

It might more reliably set the focus and there are options to activate it, set selection, set visible ranges, etc.

Reply:
I have _no idea_ how one would use this. @olegtk ?